diff -x .git -uNr redis-2.4.14/00-RELEASENOTES redis/00-RELEASENOTES
--- redis-2.4.14/00-RELEASENOTES	2012-05-23 11:32:58.000000000 +0200
+++ redis/00-RELEASENOTES	2012-06-09 18:04:57.386359000 +0200
@@ -18,7 +18,7 @@
 CHANGELOG
 ---------
 
-What's new in Redis 2.4.13
+What's new in Redis 2.4.14
 =========================
 
 UPGRADE URGENCY: high, many non trivial bugs fixed in this release.
diff -x .git -uNr redis-2.4.14/src/redis-cli.c redis/src/redis-cli.c
--- redis-2.4.14/src/redis-cli.c	2012-05-23 11:32:58.000000000 +0200
+++ redis/src/redis-cli.c	2012-06-09 18:04:57.405359000 +0200
@@ -35,6 +35,7 @@
 #include <string.h>
 #include <stdlib.h>
 #include <unistd.h>
+#include <time.h>
 #include <ctype.h>
 #include <errno.h>
 #include <sys/stat.h>
diff -x .git -uNr redis-2.4.14/src/zmalloc.h redis/src/zmalloc.h
--- redis-2.4.14/src/zmalloc.h	2012-05-23 11:32:58.000000000 +0200
+++ redis/src/zmalloc.h	2012-06-09 18:04:57.410359000 +0200
@@ -38,7 +38,7 @@
 #if defined(USE_TCMALLOC)
 #define ZMALLOC_LIB ("tcmalloc-" __xstr(TC_VERSION_MAJOR) "." __xstr(TC_VERSION_MINOR))
 #include <google/tcmalloc.h>
-#if TC_VERSION_MAJOR >= 1 && TC_VERSION_MINOR >= 6
+#if (TC_VERSION_MAJOR == 1 && TC_VERSION_MINOR >= 6) || (TC_VERSION_MAJOR > 1)
 #define HAVE_MALLOC_SIZE 1
 #define zmalloc_size(p) tc_malloc_size(p)
 #else
@@ -47,11 +47,10 @@
 
 #elif defined(USE_JEMALLOC)
 #define ZMALLOC_LIB ("jemalloc-" __xstr(JEMALLOC_VERSION_MAJOR) "." __xstr(JEMALLOC_VERSION_MINOR) "." __xstr(JEMALLOC_VERSION_BUGFIX))
-#define JEMALLOC_MANGLE
 #include <jemalloc/jemalloc.h>
-#if JEMALLOC_VERSION_MAJOR >= 2 && JEMALLOC_VERSION_MINOR >= 1
+#if (JEMALLOC_VERSION_MAJOR == 2 && JEMALLOC_VERSION_MINOR >= 1) || (JEMALLOC_VERSION_MAJOR > 2)
 #define HAVE_MALLOC_SIZE 1
-#define zmalloc_size(p) JEMALLOC_P(malloc_usable_size)(p)
+#define zmalloc_size(p) je_malloc_usable_size(p)
 #else
 #error "Newer version of jemalloc required"
 #endif
