diff -uNr nginx-0.8.29.orig/auto/lib/libatomic/conf nginx-0.8.29/auto/lib/libatomic/conf
--- nginx-0.8.29.orig/auto/lib/libatomic/conf	2009-11-27 23:00:39.000000000 +0100
+++ nginx-0.8.29/auto/lib/libatomic/conf	2009-12-04 00:07:33.009436365 +0100
@@ -14,7 +14,8 @@
     ngx_feature="atomic_ops library"
     ngx_feature_name=NGX_HAVE_LIBATOMIC
     ngx_feature_run=yes
-    ngx_feature_incs="#include <atomic_ops.h>"
+    ngx_feature_incs="#define AO_REQUIRE_CAS
+                      #include <atomic_ops.h>"
     ngx_feature_path=
     ngx_feature_libs="-latomic_ops"
     ngx_feature_test="long  n = 0;
diff -uNr nginx-0.8.29.orig/src/os/unix/ngx_atomic.h nginx-0.8.29/src/os/unix/ngx_atomic.h
--- nginx-0.8.29.orig/src/os/unix/ngx_atomic.h	2009-11-25 18:55:25.000000000 +0100
+++ nginx-0.8.29/src/os/unix/ngx_atomic.h	2009-12-04 00:07:59.839436195 +0100
@@ -14,6 +14,7 @@
 
 #if (NGX_HAVE_LIBATOMIC)
 
+#define AO_REQUIRE_CAS
 #include <atomic_ops.h>
 
 #define NGX_HAVE_ATOMIC_OPS  1
