--- src/http/ngx_http_script.c.orig	2009-09-07 12:01:26.000000000 +0200
+++ src/http/ngx_http_script.c	2009-10-15 14:10:13.000000000 +0200
@@ -1428,12 +1428,14 @@
         case ngx_http_script_file_dir:
         case ngx_http_script_file_exists:
         case ngx_http_script_file_exec:
+        case ngx_http_script_file_not_empty:
              goto false_value;
 
         case ngx_http_script_file_not_plain:
         case ngx_http_script_file_not_dir:
         case ngx_http_script_file_not_exists:
         case ngx_http_script_file_not_exec:
+        case ngx_http_script_file_empty:
              goto true_value;
         }
 
@@ -1488,6 +1490,18 @@
             goto false_value;
         }
         goto true_value;
+
+    case ngx_http_script_file_empty:
+        if (of.size == 0) {
+             goto true_value;
+        }
+        goto false_value;
+
+    case ngx_http_script_file_not_empty:
+        if (of.size > 0) {
+            goto false_value;
+        }
+        goto true_value;
     }
 
 false_value:
--- src/http/ngx_http_script.h.orig	2009-04-27 13:32:33.000000000 +0200
+++ src/http/ngx_http_script.h	2009-10-15 14:06:32.000000000 +0200
@@ -172,7 +172,9 @@
     ngx_http_script_file_exists,
     ngx_http_script_file_not_exists,
     ngx_http_script_file_exec,
-    ngx_http_script_file_not_exec
+    ngx_http_script_file_not_exec,
+    ngx_http_script_file_empty,
+    ngx_http_script_file_not_empty
 } ngx_http_script_file_op_e;
 
 
--- src/http/modules/ngx_http_rewrite_module.c.orig	2009-03-06 13:15:07.000000000 +0100
+++ src/http/modules/ngx_http_rewrite_module.c	2009-10-15 14:14:31.000000000 +0200
@@ -810,6 +810,11 @@
             return NGX_CONF_OK;
         }
 
+        if (p[1] == 'z') {
+            fop->op = ngx_http_script_file_empty;
+            return NGX_CONF_OK;
+        }
+
         if (p[0] == '!') {
             if (p[2] == 'f') {
                 fop->op = ngx_http_script_file_not_plain;
@@ -830,6 +835,11 @@
                 fop->op = ngx_http_script_file_not_exec;
                 return NGX_CONF_OK;
             }
+
+            if (p[2] == 'z') {
+                fop->op = ngx_http_script_file_not_empty;
+                return NGX_CONF_OK;
+            }
         }
 
         ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
