Index: turbojson/jsonsupport.py
===================================================================
--- turbojson/jsonsupport.py	(revision 5187)
+++ turbojson/jsonsupport.py	(working copy)
@@ -35,7 +35,10 @@
 
         """
         [info.pop(item) for item in info.copy() if (item.startswith("tg_") and item != "tg_flash")]
-        output = self.encoder.encode(info)
+        try:
+            output = self.encoder.encode(info)
+        except UnicodeDecodeError:
+            output = jsonify.encode(info)
         if isinstance(output, unicode):
             output = output.encode(self.encoding)
         return output
