--- sqlobject/col.py.orig	2007-09-05 18:20:04.000000000 +0200
+++ sqlobject/col.py	2007-09-05 18:21:01.000000000 +0200
@@ -555,9 +555,14 @@
     def from_python(self, value, state):
         if value is None:
             return None
+        if isinstance(value, unicode):
+            return value
         if isinstance(value, str):
             return value
-        return value.encode(self.db_encoding)
+        try:
+            return value.encode(self.db_encoding)
+        except AttributeError:
+            return None
 
 class SOUnicodeCol(SOStringLikeCol):
     def __init__(self, **kw):
