--- Jamroot	2009-08-23 18:17:37.000000000 +0200
+++ Jamroot.gentoo	2009-10-21 13:31:46.000000000 +0200
@@ -352,7 +352,7 @@
         if $(layout) = versioned
         {
             result = [ common.format-name
-                <base> <toolset> <threading> <runtime> -$(BOOST_VERSION_TAG)
+                <base> <threading> <runtime> -$(BOOST_VERSION_TAG)
                 -$(BUILD_ID)
                 : $(name) : $(type) : $(property-set) ] ;
         }        
