diff --git a/src/makefile b/src/makefile
index a75d5a6..40f8a46 100644
--- a/src/makefile
+++ b/src/makefile
@@ -256,21 +256,21 @@ none:
 	@echo "where PLATFORM is one of these:"
 	@echo "   $(PLATS)"
 
-all: $(SOCKET_SO) $(MIME_SO)
+all: $(SOCKET_SO) $(MIME_SO) $(UNIX_SO)
 
 $(SOCKET_SO): $(SOCKET_OBJS)
-	$(LD) $(SOCKET_OBJS) $(LDFLAGS)$@ 
+	$(LD) $(SOCKET_OBJS) $(LDFLAGS) $@ 
 
 $(MIME_SO): $(MIME_OBJS)
-	$(LD) $(MIME_OBJS) $(LDFLAGS)$@ 
+	$(LD) $(MIME_OBJS) $(LDFLAGS) $@ 
 
 all-unix: all $(UNIX_SO) $(SERIAL_SO)
 
 $(UNIX_SO): $(UNIX_OBJS)
-	$(LD) $(UNIX_OBJS) $(LDFLAGS)$@ 
+	$(LD) $(UNIX_OBJS) $(LDFLAGS) $@ 
 
 $(SERIAL_SO): $(SERIAL_OBJS)
-	$(LD) $(SERIAL_OBJS) $(LDFLAGS)$@
+	$(LD) $(SERIAL_OBJS) $(LDFLAGS) $@
 
 install: 
 	$(INSTALL_DIR) $(INSTALL_TOP_SHARE)
