Index: src/pkg/time/zoneinfo_read.go
===================================================================
--- a/src/pkg/time/zoneinfo_read.go
+++ b/src/pkg/time/zoneinfo_read.go
@@ -68,7 +68,7 @@
 
 	// 1-byte version, then 15 bytes of padding
 	var p []byte
-	if p = d.read(16); len(p) != 16 || p[0] != 0 && p[0] != '2' {
+	if p = d.read(16); len(p) != 16 || p[0] != 0 && p[0] != '2' && p[0] != '3' {
 		return nil, badData
 	}
 
@@ -123,7 +123,7 @@
 		return nil, badData
 	}
 
-	// If version == 2, the entire file repeats, this time using
+	// If version == 2 or 3, the entire file repeats, this time using
 	// 8-byte ints for txtimes and leap seconds.
 	// We won't need those until 2106.
 
Index: src/pkg/time/zoneinfo_test.go
===================================================================
--- a/src/pkg/time/zoneinfo_test.go
+++ b/src/pkg/time/zoneinfo_test.go
@@ -9,6 +9,15 @@
 	"time"
 )
 
+func TestVersion3(t *testing.T) {
+	time.ForceZipFileForTesting(true)
+	defer time.ForceZipFileForTesting(false)
+	_, err := time.LoadLocation("Asia/Jerusalem")
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
 // Test that we get the correct results for times before the first
 // transition time.  To do this we explicitly check early dates in a
 // couple of specific timezones.
