diff -uNr php-5.2.10-orig/sapi/cgi/fpm/fpm_atomic.h php-5.2.10/sapi/cgi/fpm/fpm_atomic.h
--- php-5.2.10-orig/sapi/cgi/fpm/fpm_atomic.h	2009-09-23 19:01:28.991073365 +0200
+++ php-5.2.10/sapi/cgi/fpm/fpm_atomic.h	2009-09-23 19:06:32.401068088 +0200
@@ -59,7 +59,15 @@
 
 #else
 
-#error unsupported processor. please write a patch and send it to me
+#if (__arch64__ || __arch64)
+typedef uint64_t                    atomic_uint_t;
+#else
+typedef uint32_t                    atomic_uint_t;
+#endif
+
+typedef volatile atomic_uint_t      atomic_t;
+
+#define atomic_cmp_set(a,b,c) __sync_bool_compare_and_swap(a,b,c)
 
 #endif
 
