--- dk-filter/dk-filter.c.orig	2006-07-17 16:33:44.000000000 -0700
+++ dk-filter/dk-filter.c	2006-07-17 16:35:45.000000000 -0700
@@ -2076,6 +2076,13 @@
 	dfc = cc->cctx_msg;
 	assert(dfc != NULL);
 
+	/* Determine the job ID for logging. */
+	if (dfc->mctx_jobid == 0 || strcmp(dfc->mctx_jobid, JOBIDUNKNOWN) == 0) {
+		char *jobid = smfi_getsymval(ctx, "i");
+		if (jobid != 0)
+			dfc->mctx_jobid = jobid;
+	}
+
 	/* get hostname; used in the X header and in new MIME boundaries */
 	hostname = smfi_getsymval(ctx, "j");
 	if (hostname == NULL)
