--- cdrecord-2.01/librscg/scsi-remote.c.org	2004-08-30 16:09:33.000000000 -0600
+++ cdrecord-2.01/librscg/scsi-remote.c	2004-08-30 16:11:06.000000000 -0600
@@ -1071,9 +1071,9 @@
 		/*
 		 * Become 'locuser' to tell the rsh program the local user id.
 		 */
-		if (getuid() != pw->pw_uid &&
-		    setuid(pw->pw_uid) == -1) {
-			errmsg("setuid(%lld) failed.\n",
+		if ((pw->pw_uid) != geteuid() &&
+		    seteuid(pw->pw_uid) == -1) {
+			errmsg("seteuid(%lld) failed.\n",
 							(Llong)pw->pw_uid);
 			_exit(EX_BAD);
 			/* NOTREACHED */
--- cdrtools-2.01/rscsi/rscsi.dfl.org	2005-05-07 20:19:15.930567520 +0200
+++ cdrtools-2.01/rscsi/rscsi.dfl	2005-05-07 20:19:21.072785784 +0200
@@ -11,6 +11,8 @@
 # The file where debug info should go to.
 # If you don't like debugging (e.g. for speed) comment out
 # the this line.
+# Security note: Set this to a safe place to write output, such as your home
+# directory
 #
 #DEBUG=/tmp/RSCSI
 
