--- cdrtools-2.01.01/RULES/mk-gmake.id.warnings	2005-07-22 16:01:01.000000000 +0200
+++ cdrtools-2.01.01/RULES/mk-gmake.id	2005-07-22 16:05:27.000000000 +0200
@@ -30,7 +30,7 @@
 
 ifndef		ARCH_DONE
 
-__gmake_warn:=	$(shell cat $(SRCROOT)/$(RULESDIR)/gmake.wrn 1>&2; sleep 2)
+#__gmake_warn:=	$(shell cat $(SRCROOT)/$(RULESDIR)/gmake.wrn 1>&2; sleep 2)
 
 _MACHCMD=	(mach || uname -p || true)	2> /dev/null
 _ARCHCMD=	(arch || /usr/ucb/arch || true)	2> /dev/null
--- cdrtools-2.01.01/cdrecord/cdrecord.c.warnings	2005-07-22 16:01:01.000000000 +0200
+++ cdrtools-2.01.01/cdrecord/cdrecord.c	2005-07-22 16:01:01.000000000 +0200
@@ -4894,6 +4894,8 @@
 #ifdef	HAVE_UNAME
 	struct	utsname	un;
 
+	/* Try not to break the diff for each cdrecord release */
+	return;
 	if (uname(&un) >= 0) {
 		/*
 		 * I really hope that the Linux kernel developers will soon
