--- source/inlines.cpp.orig     2009-12-29 21:08:46.400000000 +0100
+++ source/inlines.cpp  2009-12-29 21:22:16.820000000 +0100
@@ -59,7 +59,10 @@
 #include <geos/geom/CoordinateArraySequenceFactory.inl>
 #include <geos/noding/SegmentString.inl>
 #include <geos/noding/snapround/HotPixel.inl>
+// see http://trac.osgeo.org/geos/ticket/222
+#if ! defined(__ARM_EABI__)
 #include <geos/noding/snapround/MCIndexSnapRounder.inl>
+#endif // !defined __ARM_EABI__
 #include <geos/noding/MCIndexNoder.inl>

 #endif // defined __CYGWIN__
