--- cyrus-imapd-2.3.6/imap/message.c.munge8bit	2006-03-11 00:41:57.000000000 +0100
+++ cyrus-imapd-2.3.6/imap/message.c	2006-06-05 14:27:19.349694000 +0200
@@ -230,6 +230,7 @@
     int n;
     int sawcr = 0, sawnl;
     int reject8bit = config_getswitch(IMAPOPT_REJECT8BIT);
+    int munge8bit = config_getswitch(IMAPOPT_MUNGE8BIT);
     int inheader = 1, blankline = 1;
 
     while (size) {
@@ -274,7 +275,7 @@
 			/* We have been configured to reject all mail of this
 			   form. */
 			if (!r) r = IMAP_MESSAGE_CONTAINS8BIT;
-		    } else {
+		    } else if (munge8bit) {
 			/* We have been configured to munge all mail of this
 			   form. */
 			*p = 'X';
--- cyrus-imapd-2.3.6/imap/spool.c.munge8bit	2005-10-24 14:58:43.000000000 +0200
+++ cyrus-imapd-2.3.6/imap/spool.c	2006-06-05 14:27:19.349694000 +0200
@@ -140,6 +140,7 @@
     state s = NAME_START;
     int r = 0;
     int reject8bit = config_getswitch(IMAPOPT_REJECT8BIT);
+    int munge8bit = config_getswitch(IMAPOPT_MUNGE8BIT);
     const char **skip = NULL;
 
     if (namelen == 0) {
@@ -266,7 +267,7 @@
 			   form. */
 			r = IMAP_MESSAGE_CONTAINS8BIT;
 			goto ph_error;
-		    } else {
+		    } else if (munge8bit) {
 			/* We have been configured to munge all mail of this
 			   form. */
 			c = 'X';
--- cyrus-imapd-2.3.6/lib/imapoptions.munge8bit	2006-03-30 18:06:46.000000000 +0200
+++ cyrus-imapd-2.3.6/lib/imapoptions	2006-06-05 14:34:06.171118750 +0200
@@ -497,6 +497,12 @@
 { "mboxlist_db", "skiplist", STRINGLIST("flat", "berkeley", "berkeley-hash", "skiplist")}
 /* The cyrusdb backend to use for the mailbox list. */
 
+{ "munge8bit", 1, SWITCH }
+/* If enabled, lmtpd munges messages with 8-bit characters.  These characters
+   are changed to .  If \fBreject8bit\fR is enabled, setting \fBmunge8bit\fR
+   has no effect.  (A proper soultion to non-ASCII characters in headers is
+   offered by RFC 2047 and its predecessors.) */
+
 { "metapartition_files", "", BITFIELD("header", "index", "cache", "expunge", "squat") }
 /* Space-separated list of metadata files to be stored on a
    \fImetapartition\fR rather than in the mailbox directory on a spool
@@ -773,9 +779,7 @@
 
 { "reject8bit", 0, SWITCH }
 /* If enabled, lmtpd rejects messages with 8-bit characters in the
-   headers.  Otherwise, 8-bit characters are changed to `X'.  (A
-   proper solution to non-ASCII characters in headers is offered by  
-   RFC 2047 and its predecessors.) */
+   headers. */
 
 { "rfc2046_strict", 0, SWITCH }
 /* If enabled, imapd will be strict (per RFC 2046) when matching MIME
--- cyrus-imapd-2.3.6/man/deliver.8.munge8bit	2004-06-24 17:16:31.000000000 +0200
+++ cyrus-imapd-2.3.6/man/deliver.8	2006-06-05 14:27:19.349694000 +0200
@@ -147,8 +147,10 @@
 Accept messages using the LMTP protocol.
 .SH NOTES
 Depending on the setting of \fBreject8bit\fR in \fBimapd.conf(5)\fR, deliver
-either rejects messages with 8-bit-set characters in the headers or 
-changes these characters to `X'.
+either rejects/accepts messages with 8-bit-set characters in the headers.
+If we accept messages with 8-bit-set characters in the headers, then depending
+on the setting of \fBmunge8bit\fR, these characters are either left un-touched
+or changed to .
 This is because such characters can't be interpreted since the
 character set is not known, although some communities not well-served by
 US-ASCII assume that those characters can be used to represent characters not
