--- imtest/Makefile.in.orig	2007-09-07 21:45:46.000000000 +0200
+++ imtest/Makefile.in	2007-09-07 21:45:52.000000000 +0200
@@ -72,7 +72,7 @@
 all: imtest
 
 install:
-	$(INSTALL) -s -m 755 imtest $(DESTDIR)$(exec_prefix)/bin
+	$(INSTALL) -m 755 imtest $(DESTDIR)$(exec_prefix)/bin
 	ln -f $(DESTDIR)$(exec_prefix)/bin/imtest $(DESTDIR)$(exec_prefix)/bin/pop3test
 	ln -f $(DESTDIR)$(exec_prefix)/bin/imtest $(DESTDIR)$(exec_prefix)/bin/nntptest
 	ln -f $(DESTDIR)$(exec_prefix)/bin/imtest $(DESTDIR)$(exec_prefix)/bin/lmtptest
