--- configure.in.orig	2006-06-12 10:44:15.000000000 -0700
+++ configure.in	2006-06-12 10:49:00.000000000 -0700
@@ -479,7 +479,7 @@
         $with_afs=/usr/local
     fi
     CFLAGS="${CFLAGS} -I${with_afs}/include"
-    AFS_LIBS="${with_afs}/lib/afs/libkauth.a ${with_afs}/lib/afs/libprot.a  ${with_afs}/lib/afs/libauth.a ${with_afs}/lib/afs/libsys.a ${with_afs}/lib/librxkad.a ${with_afs}/lib/librx.a ${with_afs}/lib/afs/libsys.a ${with_afs}/lib/libubik.a  ${with_afs}/lib/liblwp.a ${with_afs}/lib/afs/util.a"
+    AFS_LIBS=" -lafsrpc -lafsauthent ${with_afs}/lib/afs/libkauth.a ${with_afs}/lib/afs/libprot.a  ${with_afs}/lib/afs/libauth.a ${with_afs}/lib/afs/libsys.a ${with_afs}/lib/librxkad.a ${with_afs}/lib/librx.a ${with_afs}/lib/afs/libsys.a ${with_afs}/lib/libubik.a  ${with_afs}/lib/liblwp.a ${with_afs}/lib/afs/util.a"
     if test -f ${with_afs}/lib/afs/libaudit.a; then
       AFS_LIBS="$AFS_LIBS ${with_afs}/lib/afs/libaudit.a"
     fi
