diff -uNr greenlet-0.2.orig/slp_platformselect.h greenlet-0.2/slp_platformselect.h
--- greenlet-0.2.orig/slp_platformselect.h	2010-01-19 23:28:30.496801226 +0100
+++ greenlet-0.2/slp_platformselect.h	2010-01-19 23:27:54.586800864 +0100
@@ -20,4 +20,6 @@
 #include "switch_s390_unix.h"	/* Linux/S390 zSeries (identical) */
 #elif defined(__GNUC__) && defined(__arm__) && defined(__APPLE__)
 #include "switch_arm_iphoneos_gcc.h" /* Apple IPhoneOS on arm hardware with gcc */
+#elif defined(__GNUC__) && defined(__arm__)
+#include "switch_arm32_gcc.h" /* gcc using arm32 */
 #endif
diff -uNr greenlet-0.2.orig/switch_arm32_gcc.h greenlet-0.2/switch_arm32_gcc.h
--- greenlet-0.2.orig/switch_arm32_gcc.h	1970-01-01 01:00:00.000000000 +0100
+++ greenlet-0.2/switch_arm32_gcc.h	2008-06-11 23:22:27.000000000 +0200
@@ -0,0 +1,49 @@
+/*
+ * this is the internal transfer function.
+ *
+ * HISTORY
+ * 14-Aug-06 File creation. Ported from Arm Thumb. Sylvain Baro
+ *  3-Sep-06 Commented out saving of r1-r3 (r4 already commented out) as I
+ *           read that these do not need to be saved.  Also added notes and
+ *           errors related to the frame pointer. Richard Tew.
+ *
+ * NOTES
+ *
+ *   It is not possible to detect if fp is used or not, so the supplied
+ *   switch function needs to support it, so that you can remove it if
+ *   it does not apply to you.
+ *
+ * POSSIBLE ERRORS
+ *
+ *   "fp cannot be used in asm here"
+ *
+ *   - Try commenting out "fp" in REGS_TO_SAVE.
+ *
+ */
+
+#define STACK_REFPLUS 1
+
+#ifdef SLP_EVAL
+#define STACK_MAGIC 0
+#define REGS_TO_SAVE "r4", "r5", "r6", "r7", "r8", /* "r10", "r11", */ "ip", "lr"
+
+static int
+slp_switch(void)
+{
+	register int *stackref, stsizediff;
+	__asm__ volatile ("" : : : REGS_TO_SAVE);
+	__asm__ ("mov %0,sp" : "=g" (stackref));
+	{
+		SLP_SAVE_STATE(stackref, stsizediff);
+		__asm__ volatile (
+		    "add sp,sp,%0\n"
+		    :
+		    : "r" (stsizediff)
+		    );
+		SLP_RESTORE_STATE();
+		return 0;
+	}
+	__asm__ volatile ("" : : : REGS_TO_SAVE);
+}
+
+#endif
