Index: trac/scripts/admin.py
===================================================================
--- trac/scripts/admin.py	(revision 3816)
+++ trac/scripts/admin.py	(working copy)
@@ -270,7 +270,7 @@
             seconds = int(time.time())
         else:
             for format in [self._date_format, '%x %X', '%x, %X', '%X %x',
-                           '%X, %x', '%x', '%c', '%b %d, %Y']:
+                           '%X, %x', '%x', '%c', '%b %d, %Y', '%Y-%m-%d %X']:
                 try:
                     pt = time.strptime(t, format)
                     seconds = int(time.mktime(pt))
Index: trac/Timeline.py
===================================================================
--- trac/Timeline.py	(revision 3816)
+++ trac/Timeline.py	(working copy)
@@ -100,7 +100,7 @@
         t = time.localtime()
         if req.args.has_key('from'):
             try:
-                t = time.strptime(req.args.get('from'), '%x')
+                t = time.strptime(req.args.get('from'), '%Y-%m-%d')
             except:
                 pass
 
Index: trac/util/datefmt.py
===================================================================
--- trac/util/datefmt.py	(revision 3816)
+++ trac/util/datefmt.py	(working copy)
@@ -48,7 +48,7 @@
             return '%d %s' % (r, r == 1 and unit or unit_plural)
     return ''
 
-def format_datetime(t=None, format='%x %X', gmt=False):
+def format_datetime(t=None, format='%Y-%m-%d %X', gmt=False):
     if t is None:
         t = time.time()
     if not isinstance(t, (list, tuple, time.struct_time)):
@@ -64,7 +64,7 @@
         # the above is broken on win32, e.g. we'd get '437' instead of 'cp437'
     return unicode(text, encoding, 'replace')
 
-def format_date(t=None, format='%x', gmt=False):
+def format_date(t=None, format='%Y-%m-%d', gmt=False):
     return format_datetime(t, format, gmt)
 
 def format_time(t=None, format='%X', gmt=False):
@@ -103,7 +103,7 @@
     seconds = None
     text = text.strip()
     for format in ['%x %X', '%x, %X', '%X %x', '%X, %x', '%x', '%c',
-                   '%b %d, %Y']:
+                   '%b %d, %Y', '%Y-%m-%d %X']:
         try:
             date = time.strptime(text, format)
             seconds = time.mktime(date)
