From 15b07c83049ece7bd19eb38a6c487dcc026641e6 Mon Sep 17 00:00:00 2001
From: W-Mark Kubacki <wmark@hurrikane.de>
Date: Fri, 28 Mar 2014 00:24:03 +0100
Subject: [PATCH 2/3] update ocf patches

---
 drivers/char/random.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/char/random.c b/drivers/char/random.c
index 0316403..4b333c9 100644
--- a/drivers/char/random.c
+++ b/drivers/char/random.c
@@ -996,7 +996,7 @@ void random_input_words(__u32 *buf, size_t wordcount, int ent_count)
 
 	credit_entropy_bits(&input_pool, ent_count);
 
-	DEBUG_ENT("crediting %d bits => %d\n",
+	pr_notice("crediting %d bits => %d\n",
 		  ent_count, input_pool.entropy_count);
 	/*
 	 * Wake up waiting processes if we have enough
@@ -1026,7 +1026,7 @@ int random_input_wait(void)
         /* likely we got woken up due to a signal */
 	if (count <= 0) count = random_read_wakeup_thresh; 
 
-	DEBUG_ENT("requesting %d bits from input_wait()er %d<%d\n",
+	pr_notice("requesting %d bits from input_wait()er %d<%d\n",
 		  count,
 		  input_pool.entropy_count, random_write_wakeup_thresh);
 
-- 
1.9.1

