--- tools/backup/hot-backup.py.in.orig	2003-12-16 17:41:14.000000000 +0100
+++ tools/backup/hot-backup.py.in	2005-07-06 10:44:55.000000000 +0200
@@ -34,7 +34,14 @@
 svnadmin = "@SVN_BINDIR@/svnadmin"
 
 # Number of backups to keep around (0 for "keep them all")
-num_backups = 64
+# Configurable in /etc/env.d/80hotbackup
+envvar_hb_num='SVN_HOTBACKUP_NUM_BACKUPS'
+if os.environ.has_key(envvar_hb_num):
+	num_backups = int(os.environ[envvar_hb_num])
+else:
+	num_backups = 64
+print 'Keeping up to', num_backups, 'around.'			
+
 
 ######################################################################
 # Command line arguments
