Index: kid/__init__.py
===================================================================
--- kid/__init__.py	(revision 488)
+++ kid/__init__.py	(working copy)
@@ -229,7 +229,6 @@
 
     # the serializer to use when writing output
     serializer = output_methods['xml']
-    _filters = [transform_filter]
 
     def __init__(self, *args, **kw):
         """
@@ -245,6 +244,7 @@
                 raise template_util.TemplateAttrsError(
                     "Keyword argument %r is a reserved name." % k)
         self.__dict__.update(kw)
+        self._filters = [transform_filter]
         self._layout_classes = []
 
     def write(self, file, encoding=None,
