--- datamatrix/renderer.py.old	2008-06-09 22:08:26.341732441 +0200
+++ datamatrix/renderer.py	2008-06-09 22:08:25.801701665 +0200
@@ -76,7 +76,7 @@
         # write the buffer out to an image
         img = Image.frombuffer( 'L', 
                                 (self.width*cellsize, self.height*cellsize),
-                                buff, 'raw', mode, 0, 1 )
+                                buff)
         return img
             
 
