--- ../configure.in	2006-11-11 15:58:55.000000000 +0100
+++ configure.in	2006-11-11 16:10:14.000000000 +0100
@@ -81,13 +81,14 @@
 AC_ARG_ENABLE([debug],
               [AS_HELP_STRING([--enable-debug],[build debug version without optimization [no]])],
               [debug=$enableval], 
-              [debug=no])
+              [debug="no"])
+AM_CONDITIONAL([ENABLE_DEBUG], [test x"$enable_debug" = x"yes"])
 
 # does the user want to build demo applications?
 AC_ARG_ENABLE([apps],
               [AS_HELP_STRING([--enable-apps],[build demo applications [yes]])],
               [enable_apps=$enableval],
-              [enable_apps="yes"])
+              [enable_apps="no"])
 AM_CONDITIONAL([BUILD_APPS], [test x"$enable_apps" = x"yes"])
 
 
@@ -738,7 +738,7 @@
 AM_CONDITIONAL([BUILD_PYTHON_WRAPPERS], [test x"$have_python" = "xyes"])
 
 # check for swig itself
-SWIG=""
+SWIG="no"
 if test x"$with_swig" = "xyes"; then
     AC_MSG_NOTICE([Checking for SWIG])
 
@@ -821,6 +831,7 @@
     Use dc1394 & raw1394:     ${have_dc1394}
     Use v4l:                  ${have_v4l}
     Use v4l2:                 ${have_v4l2}
+    Use gtk:		      ${have_gtk}
     
 Wrappers for other languages =========================================
     SWIG                      ${SWIG}
