From e36fdc59af3b39abc014bfe40791fa788744ba59 Mon Sep 17 00:00:00 2001
From: W-Mark Kubacki <wmark@hurrikane.de>
Date: Thu, 27 Mar 2014 23:12:24 +0100
Subject: [PATCH 2/5] reduce cache pressure and swappiness

---
 fs/dcache.c | 2 +-
 mm/vmscan.c | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/fs/dcache.c b/fs/dcache.c
index fdbe230..bd95b6b 100644
--- a/fs/dcache.c
+++ b/fs/dcache.c
@@ -79,7 +79,7 @@
  *   dentry1->d_lock
  *     dentry2->d_lock
  */
-int sysctl_vfs_cache_pressure __read_mostly = 100;
+int sysctl_vfs_cache_pressure __read_mostly = 50;
 EXPORT_SYMBOL_GPL(sysctl_vfs_cache_pressure);
 
 __cacheline_aligned_in_smp DEFINE_SEQLOCK(rename_lock);
diff --git a/mm/vmscan.c b/mm/vmscan.c
index 05e6095..80afed8 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -129,7 +129,7 @@ struct scan_control {
 /*
  * From 0 .. 100.  Higher means more swappy.
  */
-int vm_swappiness = 60;
+int vm_swappiness = 10;
 unsigned long vm_total_pages;	/* The total number of pages which the VM controls */
 
 static LIST_HEAD(shrinker_list);
-- 
1.9.1

