diff --git a/src/thot/plugins/MakoTemplating.py b/src/thot/plugins/MakoTemplating.py
index 4ba0335..2e1a093 100644
--- a/src/thot/plugins/MakoTemplating.py
+++ b/src/thot/plugins/MakoTemplating.py
@@ -18,8 +18,6 @@ class MakoTemplate(object):
                 directories=settings['template_dir'].split(','),
                 module_directory=settings['tmp_directory'] \
                                  if 'tmp_directory' in settings else None,
-                input_encoding='utf-8',
-                output_encoding='utf-8',
                 filesystem_checks=False)
 
     def e_datetimeformat(self, format='%H:%M / %d-%m-%Y'):
@@ -31,7 +29,7 @@ class MakoTemplate(object):
         try:
             return template.render(ordinal_suffix=ordinal_suffix,
                                    datetimeformat=self.e_datetimeformat,
-                                   **kwargs);
+                                   **kwargs)
         except TopLevelLookupException, e:
             raise TemplateException(e.message)
         except Exception, e:
